File: leetcode/src/rotateList.cpp
/*
 * given a list, rotate the list to right by k places, where K >= 0
 *
 * open question:
 * if k > n, what will happen ?
 *
 * test case:
 * 1->2->3->4->5, k = 2, return 4->5->1->2->3
 * */
#include "../include/preliminary.h"

class Solution{
public:
    /*
     * walking a loop, it supports case of k > n
     * */
    ListNode* rotateRight(ListNode *head, int k){
        if(!head)    return NULL;

        ListNode *l1 = head;
        int n = 1;  // nodes count of SLL

        while(l1->next){
            l1 = l1->next;
            n++;
        }
        k %= n;
        if(k == 0)    return head;

        l1->next = head;  // make it loop

        int t = n - k;
        while(t-- > 0){
            l1 = l1->next;
        }

        head = l1->next; // new head
        l1->next = NULL;  // cut the loop

        return head;
    }
};

/* unit test is in ../cpp_unittest/rotateList_unittest */

File: leetcode/java/rotateList.java
/*
 * rotate list to right by k places, where k >= 0
 *
 * test case:
 * 1->2->3->4->5
 * k = 2,  return  4->5->1->2->3
 * k = 11, return  5->1->2->3->4              
 * */
import java.io.*;
import java.util.*;

public class rotateList{
    class ListNode{
        int val;
        ListNode next;
        ListNode(int x){
            val = x;
            next = null;
        }
    }

    public ListNode rotateRight(ListNode head, int k){
        if(head == null || k==0)    return head;
        ListNode curr = head;
        int n = 1;
        while(curr.next != null){   // 1st traverse to get length of SLL
            curr = curr.next;
            ++n;
        }
        curr.next = head;        // enable loop
        int l = n - (k % n);    // in case k > n

        while(l>0){             // 2nd traverse
            curr = curr.next;
            --l;
        }
        ListNode h = curr.next;
        curr.next = null;
        return h;
    }

    public static void main(String[] args){
        return;
    }
}
